neural_lam/datastore/mllam.py

```diff
@@ -0,0 +1,355 @@
+# Standard library
+import shutil
+import warnings
+from pathlib import Path
+from typing import List
+
+# Third-party
+import cartopy.crs as ccrs
+import mllam_data_prep as mdp
+import xarray as xr
+from loguru import logger
+from numpy import ndarray
+
+# Local
+from .base import BaseCartesianDatastore, CartesianGridShape
+
+
+class MLLAMDatastore(BaseCartesianDatastore):
+    """Datastore class for the MLLAM dataset."""
+
+    def __init__(self, config_path, n_boundary_points=30, reuse_existing=True):
+        """
+        Construct a new MLLAMDatastore from the configuration file at
+        `config_path`. A boundary mask is created with `n_boundary_points`
+        boundary points. If `reuse_existing` is True, the dataset is loaded
+        from a zarr file if it exists (unless the config has been modified
+        since the zarr was created), otherwise it is created from the
+        configuration file.
+
+        Parameters
+        ----------
+        config_path : str
+            The path to the configuration file, this will be fed to the
+            `mllam_data_prep.Config.from_yaml_file` method to then call
+            `mllam_data_prep.create_dataset` to create the dataset.
+        n_boundary_points : int
+            The number of boundary points to use in the boundary mask.
+        reuse_existing : bool
+            Whether to reuse an existing dataset zarr file if it exists and its
+            creation date is newer than the configuration file.
+
+        """
+        self._config_path = Path(config_path)
+        self._root_path = self._config_path.parent
+        self._config = mdp.Config.from_yaml_file(self._config_path)
+        fp_ds = self._root_path / self._config_path.name.replace(
+            ".yaml", ".zarr"
+        )
+
+        self._ds = None
+        if reuse_existing and fp_ds.exists():
+            # check that the zarr directory is newer than the config file
+            if fp_ds.stat().st_mtime > self._config_path.stat().st_mtime:
+                self._ds = xr.open_zarr(fp_ds, consolidated=True)
+            else:
+                logger.warning(
+                    "config file has been modified since zarr was created. "
+                    "recreating dataset."
+                )
+                shutil.rmtree(fp_ds)
+
+        if self._ds is None:
+            self._ds = mdp.create_dataset(config=self._config)
```

Comment:
When thinking about how I would use this in practice, I would probably want to just run something to create this Dataarray and save it to disk, before I start any training runs. I understand that doing e.g. a 1-epoch training run would do this, but it seems a bit annoying to have to run the training to "pre-process" the dataset.

1. Could one use mllam_data_prep directly for this? To just run the pre-processing? Would that work with how this script is set up, or would this datastore start re-creating the dataarray if I did that.
2. If not (and this is something we still might want for convenience), could we add a short cli script that just creates the dataarray? I guess it should be enough that it instantiates this datastore to trigger the dataset creation and then save it to disk.

Reply:
`mdp.create_dataset` returns an `xr.Dataset` not a `xr.DataArray`. There is one data-array in the dataset for each category of data `state`, `forcing`, `static` (these are the three "outputs" in the config file). This is the same config file as the `mllam_data_prep` cli command uses. As in you can do:

```bash
python -m mllam_data_prep example.danra.yaml
```
and that will create `example.danra.zarr` 

I should also add a type annotation here to make it clear that `self._ds` is a `xr.Dataset` (I am trying to follow the `ds_`/`da_` variable name prefix that xarray uses).

Reply:
Yes, my bad, I meant each of the Dataarrays / the Dataset. I think the `ds_`/`da_` name prefix is good to use and makes things clear. But I think it is fairly easy to follow what is a da/ds already.

Ok, that's good to know! I suppose we just have to take some care when adding things to this config file (since I think that is something we will want to do, e.g. https://github.com/mllam/neural-lam/pull/66/files#r1729167635), that this will still be the case. But I guess that is mostly about making sure that mdp still accepts it if it has more options than is needed. Or rather, we might want to do something like:

```yaml
datatore_kind: mdp
mdp_config:
    schema_version: v0.2.0
    dataset_version: v0.1.0

    output:
...
```

where mdp only cares about the `mdp_config` subsection. In that case we should be a bit extra careful with how to check if the config has changed between running `python -m mllam_data_prep example.danra.yaml` and starting training. But should not be too involved, just need to check differences specifically in the `mdp_config` subsection.

Reply:
> ```yaml
> datatore_kind: mdp
> mdp_config:
>     schema_version: v0.2.0
>     dataset_version: v0.1.0
> 
>     output:
> ...
> ```

I have been thinking something similar. Introducing a sort of "global config" like this would also avoid the need to to provide the datastore kind on the command line for each CLI.

Could we put these as separate files? As in

```yaml
# config.yaml (or something like that)
datastore:
   kind: mdp
   config_path: "example.danra.yaml"
```

Reply:
Nice, yes that type of global config will definitely be useful. 

Having it as separate files is probably preferable, yes. I am thinking that we will probably want one mdp config for the boundary and another one for the "inner area". Then if we want to swap out different boundary configs but keep the same inner region it is much more convenient if you can just point to a different boundary config file in the global config rather than having to re-write the whole thing. On the other hand it's not so nice to end up with multiple `yaml` files to config each training run, but I think it's a cost we can take in this case. You will likely not change these much once they've been set up for the specific dataset.